lib/train/transports/kubernetes.rb

```diff
@@ -0,0 +1,72 @@
+# encoding: utf-8
+#
+# Copyright 2017, Noah Kantrowitz
+
+require 'mixlib/shellout'
+
+require 'train/errors'
+
+module Train::Transports
+  class SSHFailed < Train::TransportError; end
+
+  class Kubernetes < Train.plugin(1)
+    name 'kubernetes'
+
+    include_options Train::Extras::CommandWrapper
+    option :pod, required: true
+    option :container, default: nil
+    option :kubectl_path, default: 'kubectl'
+
+    def connection(state = {})
+      opts = merge_options(options, state || {})
+      validate_options(opts)
+      opts[:logger] ||= logger
+      unless @connection && @connection_opts == opts
+        @connection ||= Connection.new(opts)
+        @connection_opts = opts.dup
+      end
+      @connection
+    end
+
+    class Connection < BaseConnection
+      def os
+        @os ||= OS.new(self)
+      end
+
+      def file(path)
+        @files[path] ||= Train::File::Remote::Linux.new(self, path)
+      end
+
+      def run_command(cmd)
+        kubectl_cmd = [options[:kubectl_path], 'exec']
```

Comment:
This implementation relies on shelling out to kubectl rather than using the API directly. This limits the usage and decreases the usability of train/inspec. Is there any reason why you decided against using https://github.com/abonas/kubeclient  directly?

Reply:
A couple of reasons:

1. Stability, the APIs change a lot and kubectl/client-go will always be the first class citizen in the ecosystem.
2. Performance, Ruby network perf is not great and because we're using it as the file transport system too it would be annoying.
3. Support, Kubeclient (and pretty much all other client libs) don't actually implement the `exec` API. Most libraries just do the REST API, which here wouldn't actually help with anything. The `exec` API is weirdly undocumented but it initiates over an HTTP POST and then upgrades to SPDY (which will be replaced with websockets soon). Implementing a client for this in Ruby would be very annoying.